ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Peras/Voting/Rules.hs

```diff
@@ -0,0 +1,537 @@
+-- 1. the hard fork summary does not cover the needed point in time, or
+-- 2. Peras is not enabled at the needed point in time.
+newtype PerasQry xs a
+  = PerasQry
+  { runPerasQry ::
+      HF.Summary xs ->
+      Either HF.PastHorizonException (PerasEnabled a)
+  }
+
+instance Functor (PerasQry xs) where
+  fmap f (PerasQry g) =
+    PerasQry $ \summary ->
+      fmap (fmap f) (g summary)
+
+instance Applicative (PerasQry xs) where
+  pure x =
+    PerasQry $ \_summary ->
+      pure (pure x)
+  PerasQry f <*> PerasQry g =
+    PerasQry $ \summary -> do
+      f' <- f summary
+      g' <- g summary
+      pure (f' <*> g')
+
+instance Monad (PerasQry xs) where
+  PerasQry fa >>= g =
+    PerasQry $ \summary -> do
+      pea <- fa summary
+      case pea of
+        PerasEnabled a -> runPerasQry (g a) summary
+        NoPerasEnabled -> pure NoPerasEnabled
+
+-- Arrival slot number of a certificate
+perasCertArrivalSlot ::
+  cert ~ WithArrivalTime cert' =>
+  cert ->
+  PerasQry xs SlotNo
+perasCertArrivalSlot cert =
+  PerasQry $
+    fmap (fmap (pure . fst3)) $
+      HF.runQuery (HF.wallclockToSlot (getArrivalTime cert))
+ where
+  fst3 (x, _, _) = x
+
+-- | Slot number at the start of a Peras round
+perasRoundStart ::
+  PerasRoundNo ->
+  PerasQry xs SlotNo
+perasRoundStart roundNo =
+  PerasQry $
+    fmap (fmap (fmap fst)) $
+      HF.runQuery (HF.perasRoundNoToSlot roundNo)
+
+-- | Candidate block slot, i.e., that of the block that's at least
+-- 'blockMinSlots' (L) old from the start of the current round
+perasCandidateSlot ::
+  PerasBlockMinSlots ->
+  PerasRoundNo ->
+  PerasQry xs SlotNo
+perasCandidateSlot blockMinSlots currRoundNo = do
+  currRoundStart <- perasRoundStart currRoundNo
+  let _L = SlotNo (unPerasBlockMinSlots blockMinSlots)
+  -- NOTE: here we need make sure that the result doesn't underflow
+  if currRoundStart >= _L
+    then pure (currRoundStart - _L)
+    else pure (SlotNo 0)
+
+-- | Chain prefix leading to the candidate block.
+perasChainAtCandidateBlock ::
+  GetHeader blk =>
+  PerasBlockMinSlots ->
+  PerasRoundNo ->
+  AnchoredFragment (Header blk) ->
+  PerasQry xs (AnchoredFragment (Header blk))
+perasChainAtCandidateBlock blockMinSlots currRoundNo currChain = do
+  candidateSlot <- perasCandidateSlot blockMinSlots currRoundNo
+  pure $ fst (AF.splitAtSlot candidateSlot currChain)
+
+{-------------------------------------------------------------------------------
+  Voting interface
+-------------------------------------------------------------------------------}
+
+-- | View of the latest certificate seen by the voter
+data LatestCertSeenView cert
+  = LatestCertSeenView
+  { lcsCert :: !cert
+  -- ^ Latest certificate seen by the voter
+  , lcsArrivalSlot :: !SlotNo
+  -- ^ Slot number at which this certificate was received
+  , lcsRoundStartSlot :: !SlotNo
+  -- ^ Starting slot number of the round where this certificate was received
+  , lcsCandidateBlockExtendsCert :: !Bool
```

Comment:
I'm still not a fan of having a field of the structure that refers to something external/implicit (the candidate block). But there is probably a good reason to do so here.

Reply:
I have already added: 

```haskell
-- NOTE: note that the voting rules do not need to know _which_ candidate block
-- is being voted for, only whether it extends the block boosted by the most
-- recent certificate seen by the voter, which is provided to the rules via
-- the 'lcsCandidateBlockExtendsCert' field inside 'latestCertSeen'.
```

To the docstring of `PerasVotingView`. Would you like me to add a pointer to that docstring instead?